### PR TITLE
Add /api/core/v3/outputs/{outputId}/full route

### LINKED ISSUE #1688: Add `/api/core/v3/outputs/{outputId}/full` route

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add /api/core/v3/outputs/{outputId}/full route

## Links to any relevant issues

Closes https://github.com/iotaledger/iota-sdk/issues/1688

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Running the added example

### CODE REVIEW

File: sdk/src/types/api/core.rs [hunk 1]
```diff
@@ -525,3 +525,58 @@ pub struct UtxoChangesResponse {
...
+pub struct OutputResponse {
+    pub(crate) output: Output,
+    pub(crate) output_id_proof: OutputIdProof,
+}
+
+/// The proof of the output identifier.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OutputIdProof {
+    pub(crate) output: Output,
```
Comment (thibault-martinez) [inline sdk/src/types/api/core.rs]: The output is again in the proof ?

Comment (Thoralf-M) [reply sdk/src/types/api/core.rs]: No, don't review drafts :trollface:

File: sdk/src/types/api/core.rs [hunk 2]
```diff
@@ -525,3 +526,84 @@ pub struct UtxoChangesResponse {
...
+        )
+    }
+}
+
+/// The proof of the output commitment.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct HashableNode {
+    #[serde(rename = "type")]
+    pub kind: u8,
+    pub l: Box<OutputCommitmentProof>,
```
Comment (DaughterOfMars) [inline sdk/src/types/api/core.rs]: r should be optional, right?

Comment (Thoralf-M) [reply sdk/src/types/api/core.rs]: Why optional? Would be just a LeafHash then, no?

Comment (DaughterOfMars) [reply sdk/src/types/api/core.rs]: well one of these can be empty, iirc

Comment (Thoralf-M) [reply sdk/src/types/api/core.rs]: Don't see in which case that would be and also don't see anything optional here
https://github.com/iotaledger/iota.go/blob/2934050abf37c9ece6a35a9c5c96f87c648164b9/merklehasher/merkle_proof.go#L36

Comment (Alex6323) [reply sdk/src/types/api/core.rs]: I think Chloé is referring to the Rust POI impl in Chronicle, but that is a bit different. I think the reference implementation changed, since the test vector hash in the TIP is now different than it was back then despite being the same block ids.

**Review by Alex6323 — DISMISSED** [review]

File: sdk/src/types/api/core.rs (hunk 2 above)
Comment (DaughterOfMars) [reply sdk/src/types/api/core.rs]: still, is it not possible to have only one child?

Comment (Alex6323) [reply sdk/src/types/api/core.rs]: you can build a merkle tree over a single element...it used to be prefixed with `l` (left) though, now it's `0`:
https://github.com/iotaledger/iota.go/blob/2934050abf37c9ece6a35a9c5c96f87c648164b9/merklehasher/merkle_hasher.go

which explains why the test vector hash changed.

Comment (DaughterOfMars) [reply sdk/src/types/api/core.rs]: I see!

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/src/types/api/core.rs [hunk 3]
```diff
@@ -528,3 +529,84 @@ pub struct UtxoChangesResponse {
     pub created_outputs: Vec<OutputId>,
     pub consumed_outputs: Vec<OutputId>,
 }
+
+/// Contains the generic [`Output`] with associated [`OutputIdProof`].
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct OutputResponse {
```
Comment (thibault-martinez) [inline sdk/src/types/api/core.rs]: What's that for ?

File: sdk/src/wallet/types/mod.rs [hunk 4]
```diff
@@ -43,6 +43,8 @@ pub struct OutputData {
     pub metadata: OutputMetadata,
     /// The actual Output
     pub output: Output,
+    /// The OutputIdProof
```
Comment (thibault-martinez) [inline sdk/src/wallet/types/mod.rs]:
```suggestion
    /// The output ID proof
```

File: sdk/src/types/api/core.rs (hunk 3 above)
Comment (Thoralf-M) [reply sdk/src/types/api/core.rs]: That's for the normal outputs route, but not used yet

**Review by DaughterOfMars — DISMISSED** [review]

File: sdk/src/types/api/core.rs [hunk 5]
```diff
@@ -528,3 +529,85 @@ pub struct UtxoChangesResponse {
...
+                    serde_json::from_value::<ValueHash>(value)
+                        .map_err(|e| serde::de::Error::custom(format!("cannot deserialize value hash: {e}")))?,
+                ),
+                _ => return Err(serde::de::Error::custom("invalid output commitment proof")),
+            },
+        )
+    }
+}
+
+/// The proof of the output commitment.
```
Comment (thibault-martinez) [inline sdk/src/types/api/core.rs]: :trollface:

File: sdk/src/wallet/types/mod.rs [hunk 6]
```diff
@@ -43,6 +43,8 @@ pub struct OutputData {
     pub metadata: OutputMetadata,
     /// The actual Output
     pub output: Output,
+    /// The output ID proof
+    pub output_id_proof: OutputIdProof,
```
Comment (thibault-martinez) [inline sdk/src/wallet/types/mod.rs:46]: Do we actually need it here?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Common Feature Sets failing

File: sdk/src/wallet/types/mod.rs (hunk 6 above)
Comment (Thoralf-M) [reply sdk/src/wallet/types/mod.rs:46]: Not sure, might be useful before signing so one can verify again that the inputs are actually what one thinks

Comment (Thoralf-M) [reply sdk/src/wallet/types/mod.rs:46]: Also thought that we could just have `OutputWithMetadata` here instead of the fields separated

**Comment by Thoralf-M** [comment]
> Common Feature Sets failing

There is a small problem that asks for changes in 55 files :trollface: 
https://github.com/Thoralf-M/iota-sdk/compare/outputs-full-route...Thoralf-M:iota-sdk:small-change
In this PR here I only modified OutputMetadata, but we also had a OutputMetadataResponse which is now redundant
And then is the question if we want to store the OutputIdProof everywhere or if we say we don't need it

File: sdk/src/types/block/output/mod.rs [hunk 7]
```diff
@@ -75,17 +78,27 @@ pub enum OutputBuilderAmount {
...
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(rename_all = "camelCase")
+)]
 pub struct OutputWithMetadata {
-    pub(crate) output: Output,
-    pub(crate) metadata: OutputMetadata,
+    pub output: Output,
+    pub output_id_proof: OutputIdProof,
+    pub metadata: OutputMetadata,
```
Comment (Alex6323) [inline sdk/src/types/block/output/mod.rs:90]: `OutputWithProofAndMetadata`? :trollface:

Comment (Alex6323) [reply sdk/src/types/block/output/mod.rs:90]: out of interest: why couldn't we put the proof inside of the metadata from a TIP perspective, or ... if we wanted to keep it separate have a distinct API call to query the proof?

File: sdk/src/wallet/types/mod.rs (hunk 6 above)
Comment (Alex6323) [reply sdk/src/wallet/types/mod.rs:46]: can you remined me? Why do we store the `output_id` in the `OutputData` if it's already in the also contained metadata?

File: sdk/src/types/block/output/mod.rs (hunk 7 above)
Comment (Thoralf-M) [reply sdk/src/types/block/output/mod.rs:90]: One could have also structured it in a different way, but it made more sense this way because the output and the proof is static data and the metadata is dynamic, also the proof can only be used together with the output, having it in the metadata without the output itself doesn't help (the proof is also returned in the route with the output alone)

**Review by Alex6323 — DISMISSED** [review]

File: sdk/src/types/block/output/mod.rs (hunk 7 above)
Comment (Alex6323) [reply sdk/src/types/block/output/mod.rs:90]: I see.

File: sdk/src/wallet/types/mod.rs (hunk 6 above)
Comment (Thoralf-M) [reply sdk/src/wallet/types/mod.rs:46]: It was added so it's available without potential error from calculating it (because in the old metadata it needed to be created from tx id and output index which could fail)
Now if we trust the metadata we could remove it I guess, but if we compute it from the output with proof, then this could still fail 
We could maybe have a builder for the output data which compares the output id from the proof with the one from the metadata :thinking:

Comment (Thoralf-M) [reply sdk/src/wallet/types/mod.rs:46]: Opened an issue https://github.com/iotaledger/iota-sdk/issues/1759

**Comment by thibault-martinez** [comment]
> > Common Feature Sets failing
> 
> There is a small problem that asks for changes in 55 files :trollface: [Thoralf-M/iota-sdk@outputs-full-route...Thoralf-M:iota-sdk:small-change](https://github.com/Thoralf-M/iota-sdk/compare/outputs-full-route...Thoralf-M:iota-sdk:small-change) In this PR here I only modified OutputMetadata, but we also had a OutputMetadataResponse which is now redundant And then is the question if we want to store the OutputIdProof everywhere or if we say we don't need it

I don't understand your answer. This is unrelated to the CI failing?

**Comment by Thoralf-M** [comment]
@thibault-martinez CI is failing because I imported an API type to types module which is not no_std compatible, so I moved the OutputWithMetadata out of the types and replaced the OutputWithMetadataResponse
And because OutputWithMetadata is used all over the place, it all needed to be updated

**Review by thibault-martinez — DISMISSED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]
